chainer/functions/activation/relu.py

```diff
@@ -12,48 +12,103 @@
     _mode = cudnn.cudnn.CUDNN_ACTIVATION_RELU
 
 
-class ReLU(function.Function):
+class ReLU(function_node.FunctionNode):
 
     """Rectified Linear Unit."""
     # TODO(beam2d): Implement in-place version.
 
+    _use_cudnn = False
+
     def check_type_forward(self, in_types):
         type_check.expect(
             in_types.size() == 1,
             in_types[0].dtype.kind == 'f',
         )
 
     def forward_cpu(self, x):
-        self.retain_inputs(())
         self.retain_outputs((0,))
         return utils.force_array(numpy.maximum(x[0], 0, dtype=x[0].dtype)),
 
     def forward_gpu(self, x):
         if chainer.should_use_cudnn('==always') and x[0].flags.c_contiguous:
+            # cupy.activation_backward requires the input.
+            # So, we retain it for backward computation.
+            self.retain_inputs((0,))
             self._use_cudnn = True
             y = cudnn.activation_forward(x[0], _mode)
         else:
-            self.retain_inputs(())
-            self._use_cudnn = False
             y = cuda.cupy.maximum(x[0], 0)
         self.retain_outputs((0,))
         return y,
 
-    def backward_cpu(self, x, gy):
-        y = self.output_data[0]
-        return utils.force_array(gy[0] * (y > 0)),
-
-    def backward_gpu(self, x, gy):
-        y = self.output_data[0]
+    def backward(self, indexes, gy):
+        y = self.get_retained_outputs()[0]
         if chainer.should_use_cudnn('==always') and self._use_cudnn:
-            gx = cudnn.activation_backward(x[0], y, gy[0], _mode)
+            x = self.get_retained_inputs()[0]
+            return ReLUGrad3().apply((x, y, gy[0]))
         else:
-            gx = cuda.elementwise(
-                'T y, T gy', 'T gx',
-                'gx = y > 0 ? gy : (T)0',
-                'relu_bwd')(y, gy[0])
+            return ReLUGrad2().apply((y, gy[0]))
+
+
+def _heaviside(x):
+    return utils.force_array((x.data > 0).astype(x.dtype))
+
+
+class ReLUGrad2(function_node.FunctionNode):
+
+    def forward_cpu(self, inputs):
+        self.retain_inputs((0,))
+        b, c = inputs
+        y = (b > 0) * c
+        return utils.force_array(y, dtype=y.dtype),
+
+    def forward_gpu(self, inputs):
+        self.retain_inputs((0,))
+        b, c = inputs
+        gx = cuda.elementwise(
+            'T y, T gy', 'T gx',
+            'gx = y > 0 ? gy : (T)0',
+            'relu_bwd')(b, c)
         return gx,
 
+    def backward(self, indexes, gy):
+        ret = []
+        if 0 in indexes:
+            ret.append(None)
```

Comment:
Could you remove the first argument from inputs and instead pass it as an argument of `__init__`? (It will simplifies the backprop, which is good for performance)